pjmedia/src/pjmedia/av_sync.c

```diff
@@ -0,0 +1,410 @@
+ * media.
+ */
+PJ_DEF(pj_status_t) pjmedia_av_sync_update_ref(
+                                pjmedia_av_sync_media* media,
+                                const pj_timestamp* ntp,
+                                const pj_timestamp* ts)
+{
+    PJ_ASSERT_RETURN(media && ntp && ts, PJ_EINVAL);
+
+    media->ref_ntp = *ntp;
+    media->ref_ts  = *ts;
+    media->is_ref_set = PJ_TRUE;
+    TRACE_((media->av_sync->pool->obj_name, "%s updates ref ntp=%ull ts=%ull",
+            media->name, ntp->u64, ts->u64));
+
+    return PJ_SUCCESS;
+}
+
+
+PJ_DEF(pj_status_t) pjmedia_av_sync_update_pts(
+                                pjmedia_av_sync_media *media,
+                                const pj_timestamp *pts,
+                                pj_int32_t *adjust_delay)
+{
+    pjmedia_av_sync *avs;
+    pj_int32_t diff;
+    pj_timestamp max_ntp;
+
+    PJ_ASSERT_RETURN(media && media->av_sync && pts, PJ_EINVAL);
+
+    /* Reset the adjustment delay */
+    if (adjust_delay)
+        *adjust_delay = 0;
+
+    /* Make sure we have a reference */
+    if (!media->is_ref_set)
+        return PJ_EINVALIDOP;
+
+    avs = media->av_sync;
+    diff = pj_timestamp_diff32(&media->ref_ts, pts);
+
+    /* Only process if:
+     * - pts is increasing, and
+     * - not jumping too far (< one minutes).
+     */
+    if (diff <= 0 || diff >= (int)media->clock_rate*60) {
+        /* Reset reference */
+        //media->is_ref_set = PJ_FALSE;
+        return (diff<=0? PJ_ETOOSMALL : PJ_ETOOBIG);
+    }
+
+    /* Update last presentation time */
+    media->last_ntp = media->ref_ntp;
+    ntp_add_ts(&media->last_ntp, diff, media->clock_rate);
+
+    /* Get NTP timestamp of the earliest media */
+    pj_grp_lock_acquire(avs->grp_lock);
+    max_ntp = avs->max_ntp;
+    pj_grp_lock_release(avs->grp_lock);
+
+    /* Check if this media is the fastest/earliest */
+    if (pj_cmp_timestamp(&media->last_ntp, &max_ntp) > 0) {
+        /* Yes, it is the fastest, update the max timestamp */
+        pj_grp_lock_acquire(avs->grp_lock);
+        avs->max_ntp = media->last_ntp;
+        pj_grp_lock_release(avs->grp_lock);
+
+        /* Check if there is any request to slow down */
+        if (avs->slowdown_req_ms) {
+            media->last_adj_delay_req = avs->slowdown_req_ms;
+            media->adj_delay_req_cnt = 0;
+            avs->slowdown_req_ms = 0;
+            TRACE_((avs->pool->obj_name, "%s is requested to slow down %dms",
+                    media->name, media->last_adj_delay_req));
+            if (adjust_delay)
+                *adjust_delay = media->last_adj_delay_req;
+
+            return PJ_SUCCESS;
+        }
+    } else {
+        /* Not the fastest. */
+        pj_timestamp ntp_diff = max_ntp;
+        unsigned ms_diff;
+
+        /* First, check the delay from the fastest. */
+        pj_sub_timestamp(&ntp_diff, &media->last_ntp);
+        ms_diff = ntp_to_ms(&ntp_diff);
+
+        /* Make sure the delay is sensible, e.g: not exceeding 60s */
```

Comment:
1 minute seems super long, even 3-5 sec is already considered long in today's standard IMO.

And shouldn't the return status be ETOOBIG?

Reply:
The delay/`ms_diff` is the lateness from the fastest, which eventually will trigger a speed-up request. If say the check is set to 5 seconds, then when a video is lagging 6 seconds behind the audio, speed-up request won't be triggered. Originally I doubt whether such boundary check is necessary.

When the media is lagging behind, it means that the supplied `pts` is too small (from the function caller PoV)?